<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [2]:
import pandas as pd
import psycopg2
import urllib.parse as req
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

# Этот блок кода можно удалить

DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [ ]:
# Этот блок кода можно удалить

connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [3]:
# Вместо предыдущих блоков можно использовать этот
# После того как все outputs будут отображены, параметры внутри create_engine
# можно удалить, потом сохранить ноутбук и загрузить его на GitHub

from sqlalchemy import create_engine

connection = create_engine("postgresql+psycopg2://skillfactory:cCkxxLVrDE8EbvjueeMedPKt@84.201.134.129:5432/project_sql")
connection.connect()

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [31]:
# Нумерацию в названии переменных проставляю в соответсвии с номером задания, 
# чтоб было проще ориентироваться

# Дальше есть множество вариантов выводов
# Вариант 1
query_3_1 = f'''
select
    count(id) as "Количество вакансий"
from 
    vacancies
'''

display(pd.read_sql_query(query_3_1, connection))

Количество вакансий
0                49197

In [34]:
# Вариант 2
display(f'Количество вакансий в базе: {pd.read_sql_query(query_3_1, connection).loc[0][0]}')

'Количество вакансий в базе: 49197'

In [42]:
# Вариант 3
from sqlalchemy import text 
results = connection.execute(query_3_1)
for record in results: 
    print(record[0])

49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [ ]:
# Общие данные получим с помощью подсчета id работодателей
query_3_2 = f'''
SELECT
    count(id)
FROM
    employers
'''

In [ ]:
# Выводим ответ с помощью f-строк
display(f'Количество работодателей в базе: {int(pd.read_sql_query(query_3_2, connection).iloc[0])}')

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [ ]:
# Считаем количество id регионов
query_3_3 = f'''
SELECT
    count(id)
FROM 
    areas
'''

In [ ]:
# Выводим ответ с помощью f-строк
display(f'Количество регионов: {int(pd.read_sql_query(query_3_3, connection).iloc[0])}')

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [ ]:
# Считаем количество сфер деятельности по количеву их id
query_3_4 = f'''
SELECT
    count(id)
FROM
    industries
'''

In [ ]:
# Выводим ответ с помощью f-строк
display(f'Количество сфер деятельности: {int(pd.read_sql_query(query_3_4, connection).iloc[0])}')

***

### <center> ***Выводы по предварительному анализу данных***
Удалось подключиться к БД, в следствии чего была получена следующая информация:
    
- Всего в базе данных нахождится 49197 вакансий;
- Работодателей в базе 23501;
- Количество охватываемых регионов - 1362;
- В базе есть вакансии из 294 разных сфер деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [ ]:
# В запросе устанавливаем LIMIT, 
# так как поусловию задания необходимо вывести только ТОП-5 регионов
query_4_1 = f'''
SELECT
    a.name,
    count(a.id)
FROM 
    vacancies v
JOIN areas a on a.id=v.area_id
GROUP BY a.name
ORDER BY 2 desc
LIMIT 5
                    
'''


In [ ]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
display('ТОП - 5 регионов по количеству вакансий')
df_4_1

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
# Используем пустые значения в фильтрации
query_4_2 = f'''
SELECT
    count(id)
FROM 
    vacancies
WHERE (salary_from is not null) or (salary_to is not null)
                    
'''

In [ ]:
# Выводим ответ с помощью f-строк
display(f'Количество вакансий, где заполнено хотя бы одно поле с зарплатой: {int(pd.read_sql_query(query_4_2, connection).iloc[0])}')

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [ ]:
# При выводе среднего значения сразу округляем его
query_4_3 = f'''
SELECT
    round(avg(salary_from)) salary_from,
    round(avg(salary_to)) salary_to
FROM
    vacancies
'''

In [ ]:
df_4_3 = pd.read_sql_query(query_4_3, connection)
display('Средние значения для верхней и нижней границ зарпалтной вилки')
df_4_3

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
# Для получения ТОЛЬКО второго значения устанавливаем OFFSET и LIMIT
query_4_4 = f'''
SELECT
    schedule,
    employment,
    count(id)
FROM
    vacancies
GROUP BY schedule, employment
ORDER BY 3 desc
OFFSET 1
LIMIT 1
'''

In [ ]:
df_4_4 = pd.read_sql_query(query_4_4, connection)
display('Вторая строка из запроса')
df_4_4

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [ ]:
# Так как в запросе присутствует агрегатная функция, использем группировку по опыту работы
query_4_5 = f'''
SELECT
    experience as experience,
    count(id) as cnt
FROM 
    vacancies
GROUP BY experience
ORDER BY 2
'''

In [ ]:
df_4_5 = pd.read_sql_query(query_4_5, connection)
display('Требуемый опыт работы в порядке убывания вакансий')
df_4_5

***

### <center> ***Вывод по детальному анализу вакансий:***

В ходе детального анализа ваконсий выяснили, что самое большое количество вакансий размещено, как и ожидалось, в областных городах. 
Далее узнали, что средняя минимальная заработная плата составляет 71065₽, а верхняя граница находится на уровне	110537₽. Разница между этими показателями составляет порядка 36%, что обусловлено, скорее всего, более высокой стоимостью жизни в крупных городах.
Самый распространенный тип занятости (16% от общего числа вакансий) - "удаленная работа с полной занятостью". 
Большая часть вакансий в требованиях к соискателю содержит "Опыт работы от 1 до 3 лет", таких вакансий 53%. Но для тех соискателей, у которых нет опыта работы, тоже есть вакансии, количество которых составляет 14,5% от общего числа вакансий. 

In [ ]:
df_4_5['cnt'].astype(float)

vaccine_combinations = df_4_5
fig = plt.figure(figsize=(3.5, 3.5))
axes = fig.add_axes([0,0,1,1])
axes.pie(
    df_4_5['cnt'],
    labels=df_4_5['experience'],
    autopct='%.1f%%',
    explode = [0, 0, 0, 0]
)
axes.set_title('Доля вакансий от опыта для DS')

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [ ]:
# Объединяем 2 таблицы и ограничеваем вывод через LIMIT

query_5_1 = f'''
SELECT 
    e.name,
    count(v.id)
FROM employers e
LEFT JOIN vacancies v on v.employer_id=e.id
GROUP BY e.name
ORDER BY 2 desc
LIMIT 5
'''

In [ ]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
display(f'Первое место: {df_5_1.iloc[0][0]}; Пятое место: {df_5_1.iloc[4][0]}')
display(df_5_1)

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
# Так как нужно узнать один регион с наибольшим количеством вакансий, 
# ограничиваем вывод с помощью LIMIT'а
query_5_2 = f'''
SELECT 
    a.name,
    count(distinct e.id) employers, 
    count(distinct v.id) vacancies
FROM areas a
LEFT JOIN vacancies v on v.area_id=a.id
LEFT JOIN employers e on a.id=e.area
WHERE v.id is NULL
    GROUP BY a.name
    ORDER BY employers desc
    LIMIT 1
'''


In [ ]:
# результат запроса
df_5_2 = pd.read_sql_query(query_5_2, connection)
display(f'Регион с наибольшим количеством работодателей ({df_5_2.iloc[0][1]}), но с  отсутствующими вакансиями: {df_5_2.iloc[0][0]}')

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [ ]:
# Ичпользуем группировку и ограничиваем вывод
query_5_3 = f'''
SELECT 
    e.name,
    count(distinct v.area_id) areas
FROM vacancies v
JOIN employers e on e.id=v.employer_id
    GROUP BY e.name
    ORDER BY areas desc
    LIMIT 1
'''

In [ ]:
df_5_3 = pd.read_sql_query(query_5_3, connection)
display(f'Максимальное количество регионов для одного работодателя: {df_5_3.iloc[0][1]}')


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [ ]:
# Используем LEFT JOIN, от оставит пропуски в строках без совпадений
query_5_4 = f'''
SELECT 
    count(e.name)
FROM employers e
LEFT JOIN employers_industries ei on e.id=ei.employer_id
LEFT JOIN industries i on i.id=ei.industry_id
    WHERE industry_id is null
'''

In [ ]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)

display(f'Количество работадателей, у которых не указана сфера деятельности. : {int(df_5_4.iloc[0])}')

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [ ]:
# Используем HAVING, так как есть агрегатные функции в запросе с которыми WHERE не работает
query_5_5 = f'''
SELECT 
    e.name,
    count(i.id)
FROM employers e
LEFT JOIN employers_industries ei on e.id=ei.employer_id
LEFT JOIN industries i on i.id=ei.industry_id
    GROUP BY e.name
    HAVING count(i.id) = 4
    ORDER BY 1
    OFFSET 2
    LIMIT 1
'''

In [ ]:
df_5_5 = pd.read_sql_query(query_5_5, connection)
display('Название компании, находящейся на 3-м месте, среди компаний с 4-мя сферами деятельности : {}'.format(df_5_5.iloc[0]['name']))

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [ ]:
# текст запроса
query_5_6 = f'''
SELECT 
    count(ei.employer_id) cnt
FROM EMPLOYERS_INDUSTRIES ei
JOIN INDUSTRIES i on i.id = ei.industry_id
    WHERE i.name = 'Разработка программного обеспечения'
'''

In [ ]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
display(f'Количество работодателей, у которых  в качестве сферы деятельности указана «Разработка программного обеспечения»: {int(df_5_6.iloc[0])}')

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [ ]:
# код для получения списка городов-милионников
path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
cities = tuple(pd.read_html(url)[1]['Город'])

In [ ]:
#  Добавим Total через union ALL и аналогичный запрос
query_5_7 = f'''
(SELECT 
    a.name,
    count(*) cnt
FROM employers e
JOIN vacancies v on e.id=v.employer_id
JOIN areas a on a.id=v.area_id
WHERE e.name = 'Яндекс' and a.name in {cities}
GROUP BY a.id
ORDER BY 2)

UNION ALL

(SELECT 'Total',
   	count(*) cnt
FROM employers e
JOIN vacancies v on e.id = v.employer_id
JOIN areas a on a.id = v.area_id
WHERE e.name = 'Яндекс' and a.name in {cities})
'''

In [ ]:
df_5_7 = pd.read_sql_query(query_5_7, connection)
display('В выборке получилось {} строк'.format(df_5_7.shape[0]))
display('В строке "Total" получился результат {}'.format(int(df_5_7[df_5_7['name'] == 'Total']['cnt'])))
display(df_5_7)

***

### <center> ***Вывод по анализу работодателей:***
В ходе исследования мы узнали, что в ТОП-5 работодателей по количеству вакансий входят IT-копании и нефтяные компании (Яндекс - Газпром нефть).
Далее определили регион, где есть зарегистрированные работодатели, но у них нет вакансий. Так как эти регионом оказалась россия, то можно предположить, что работодатели указывают регион более точно, причем есть компания, у которой вакансии размещены в 181 регионе.
Около 1/3 работодателей не указывают сферу своей деятельности, но есть и такие компании, которые указывают несколько сфер деятельности. 
Около 20% работодателей указали сферу своей деятельности как "Разработка программного обеспечения". Это может говорить отом, что сейчас в этой отрасли наблюдается дефицит сотрудников. 


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
# Для игнорирования регистра используем ILIKE вместо LIKE
query_6_1 = f'''
SELECT 
    count(*)
FROM vacancies
WHERE name ILIKE '%data%' or name ILIKE '%данн%'
'''

In [ ]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
display(f'Количество вакансий, имеющих отношение к данным - {int(df_6_1.iloc[0])}')

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
# используем ILIKE для игнорирования регистра кроме везде, кроме ML и HTML
# (исключим возможность вхождения символов в слова, например Tea%mL%ead)
query_6_2 = f'''
SELECT 
    count(*)
FROM vacancies
WHERE (name ILIKE '%data scientist%' or
      name ILIKE '%data science%' or
      name ILIKE '%исследователь данных%' or
     (name LIKE '%ML%' and
     name NOT LIKE '%HTML%') or
      name ILIKE '%machine learning%' or
      name ILIKE '%машинн%обучен%') 
      and
     (name ILIKE '%junior%' or
      experience ILIKE '%Нет опыта%' or
      employment ILIKE '%Стажировка%')
'''


In [ ]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
display(f'Количество вакансий: {df_6_2.iloc[0][0]}')

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# Запрос, налогичный предыдущему за исключением требований к Junior-специалисту, 
# но с добавлением ключевых навыков
query_6_3 = f'''
SELECT 
    count(*)
FROM vacancies
WHERE (name ILIKE '%data scientist%' or
      name ILIKE '%data science%' or
      name ILIKE '%исследователь данных%' or
     (name LIKE '%ML%' and
      name NOT LIKE '%HTML%') or
      name ILIKE '%machine learning%' or
      name ILIKE '%машинн%обучен%') 
      and
     (key_skills ILIKE '%Python%' or
      key_skills ilILIKEike '%postgres%')
'''

In [ ]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_2, connection)
display(f'Количество вакансий, удовлетворяющих критериям: {df_6_3.iloc[0][0]}')

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# Изменяем навыки
query_6_4 = f'''
SELECT 
    count(*)
FROM vacancies
WHERE (name ILIKE '%data scientist%' or
      name ILIKE '%data science%' or
      name ILIKE '%исследователь данных%' or
     (name LIKE '%ML%' and
      name NOT LIKE '%HTML%') or
      name ILIKE '%machine learning%' or
      name ILIKE '%машинн%обучен%') 
      and
     (key_skills ILIKE '%Python%')
'''

In [ ]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
display(f'Количество вакансий для специалиста по DS со знанием Python: {df_6_4.iloc[0][0]}')

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# Считаем по количеству разделителей
query_6_5 = f'''
SELECT 
   round(avg(length(key_skills)-length(replace(key_skills, CHR(9), ''))+1), 2)
FROM vacancies
WHERE (name ILIKE '%data scientist%' or
      name ILIKE '%data science%' or
      name ILIKE '%исследователь данных%' or
     (name LIKE '%ML%' and
      name NOT LIKE '%HTML%') or
      name ILIKE '%machine learning%' or
      name ILIKE '%машинн%обучен%')
'''

In [ ]:
df_6_5 = pd.read_sql_query(query_6_5, connection)
display(f'В среднем в вакансиях для DS указывают {float(df_6_5.iloc[0])} навыков')

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
# Запрос аналогичен предыдущему, за исключением пропусков
query_6_6 = f'''
SELECT 
   	experience as experience,
   	round(avg(coalesce((salary_from + salary_to) / 2, salary_from, salary_to)), 0) as avg_ds
FROM VACANCIES
WHERE 	(name ILIKE '%data scientist%' or 
		name ILIKE '%data science%' or 
		name ILIKE '%исследователь данных%' or 
		(name LIKE '%ML%' and 
		name NOT LIKE '%HTML%') or
		name ILIKE '%machine learning%' or 
		name ILIKE '%машинн%обучен%') and  
        (salary_from is not NULL or salary_to is not NULL)
GROUP BY experience
'''

In [ ]:
df_6_6 = pd.read_sql_query(query_6_6, connection)
display('Средняя з/п для специалиста по DataScience с опытом работы от 3 до 6 лет составляет {} ₽'.format(int(
    df_6_6[df_6_6['experience'] == 'От 3 до 6 лет']['avg_ds']))
    )

***

### <center> ***Вывод по предметному анализу:***
Во время проведения предметного анализа узнали, что количество вакансий, имеющих отношение к данным (исходя из названия вакансии) составляет 1771, что составляет 3.4% от общего количества вакансий. 51 вакансия относится к начинающим специалистам, так как не требуют опыта работы.
Выяснили, что количество вакансий для специалистов DataSience, в которых указано в качестве ключевого навыка SQL и Postgres, составляет 201. Более популярное требование - знание Python. Вакансий, содержащих данное требование - 351. 
Далее выяснили, что среднее количество навыков, которые указывают работодатели в вакансиях составляет 6.4. При таких требованиях специалист, имеющий опыт работы от 3 до 6 лет,  может расчитывать на среднюю зарплату около 245.000₽.

## <center> ***Общий вывод по проекту***

Проведя анализ вакансий, мы видем, что важными критериями являются опыт работы иключевые навыки соискателей. Для соискателей без опыта работы работодатели предлагают самый низкий уровень заработной платы в DS, что и логично. Но с приобретением различных ключевых навыков и опыта работы уровень заработной платы сильно растет.

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

fig, axes= plt.subplots(1, 2, figsize=(15, 5))
sns.lineplot(x='experience', y='avg_ds', data=df_6_6, ax=axes[0])
sns.barplot(x='experience', y='cnt', data=df_4_5, order=['Нет опыта', 'От 1 года до 3 лет', 'От 3 до 6 лет', 'Более 6 лет'], ax=axes[1])
axes[0].set_title('График зависимости средней з/п от опыта для DS')
axes[1].set_title('Количество вакансий по опыту работы')
axes[0].grid(True)                             

Данные графики показывают зависимость уровня заработной платы от опыта работы, а так же количество вакансий по опыту работы. Здесь наглядно видно, что заработная плата растет по мере приобретения опыта работы в сфере, а так же видим, что самый большой спрос на специалистов с опытом работы от 1 до 3 лет.

In [ ]:
query_7_1 = f'''
SELECT 
   	experience as experience,
   	round(avg(coalesce((salary_from + salary_to) / 2, salary_from, salary_to)), 0) as avg_no_DS
FROM VACANCIES
WHERE 	(name NOT ILIKE '%data scientist%' or 
		name NOT ILIKE '%data science%' or 
		name NOT ILIKE '%исследователь данных%' or 
		(name NOT LIKE '%ML%' and 
		name NOT LIKE '%HTML%') or
		name NOT ILIKE '%machine learning%' or 
		name NOT ILIKE '%машинн%обучен%') and  
        (salary_from is not NULL or salary_to is not NULL)
GROUP BY experience
'''


In [ ]:
df_7_1 = df_6_6.merge(pd.read_sql_query(query_7_1, connection), how='left')
df_7_1['delta'] = round(df_7_1['avg_ds'] / df_7_1['avg_no_ds'] *100 - 100)
df_7_1

Эта таблица наглядно показывает, что при одинаковом количества опыта работы, в вакансиях для DS показатель заработной платы выше на 80%, чем в прочих вакансиях.
Такая разница в размере заработной платы обусловлена тем, что количество данных растет быстрее, чем "появлются" новые специалисты по работе с ними, которая требует не мало занинй, умений и имеются дяволно высокие требования. 

In [ ]:
#Закрываем соединение
connection.close()